File: reservation-frontend/src/app/core/models/reservations/reservation.ts
import { StatutReservation } from './statut-reservation.enum';
import { Client } from '../clients/client';
import { Salle } from '../salle/salle';
import { Service } from '../service/service';
import { TarificationSalle } from '../salle/tarification-salle';
import { TypePeriode } from '../salle/TypePeriode.enum';
import { Contrat } from '../contrats/contrat';
import { Paiement } from '../paiement/paiement';
import { User } from '../user/user';

export class Reservation {
  reservationId?: number;
  numeroReservation!: string;
  date!: string; // Format YYYY-MM-DD
  statut!: StatutReservation;
  montantAPayer?: number; // Corrigé de montantTotal à montantAPayer
  tokenReservation?: string;
  notes?: string;

  // Relations
  client!: Client;
  salle!: Salle;
  services: Service[] = [];
  tarificationsAppliquees: TarificationSalle[] = [];
  tarificationAppliquee?: TarificationSalle;
  typePeriode?: TypePeriode;
  contrat?: Contrat;
  paiements: Paiement[] = [];
  creePar?: User; // Ajouté pour correspondre au Java
  modifiePar?: User; // Ajouté pour correspondre au Java

  // Métadonnées système
  dateCreation?: string;
  dateModification?: string;

  constructor(
    reservationId?: number,
    numeroReservation?: string,
    date?: string,
    statut?: StatutReservation,
    montantAPayer?: number,
    client?: Client,
    salle?: Salle,
    services: Service[] = [],
    notes?: string,
    tokenReservation?: string,
    tarificationsAppliquees: TarificationSalle[] = [],
    contrat?: Contrat,
    paiements: Paiement[] = [],
    creePar?: User,
    modifiePar?: User,
    dateCreation?: string,
    dateModification?: string
  ) {
    if (reservationId !== undefined) this.reservationId = reservationId;
    if (numeroReservation) this.numeroReservation = numeroReservation;
    if (date) this.date = date;
    if (statut) this.statut = statut;
    if (montantAPayer !== undefined) this.montantAPayer = montantAPayer;
    if (client) this.client = client;
    if (salle) this.salle = salle;
    this.services = services;
    if (notes) this.notes = notes;
    if (tokenReservation) this.tokenReservation = tokenReservation;
    this.tarificationsAppliquees = tarificationsAppliquees;
    if (contrat) this.contrat = contrat;
    this.paiements = paiements;
    if (creePar) this.creePar = creePar;
    if (modifiePar) this.modifiePar = modifiePar;
    if (dateCreation) this.dateCreation = dateCreation;
    if (dateModification) this.dateModification = dateModification;
  }
}

File: Reservation/src/main/java/net/travel/reservation/entites/Reservation.java
package net.travel.reservation.entites;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@Entity
@Table(name = "reservations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long reservationId;

    @Column(nullable = false, unique = true)
    private String numeroReservation;

    @Column(nullable = false)
    private LocalDate date;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private StatutReservation statut;

    @Column(nullable = true, precision = 10, scale = 2)
    private BigDecimal montantAPayer;

    @Column(unique = true)
    private String tokenReservation;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @PrePersist
    public void prePersist() {
        this.dateCreation = LocalDateTime.now();
        if (this.tokenReservation == null) {
            this.tokenReservation = UUID.randomUUID().toString();
        }
    }

    @PreUpdate
    public void preUpdate() {
        this.dateModification = LocalDateTime.now();
    }

    // --- RELATIONS ---

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false)
    private Client client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "salle_id", nullable = false)
    private Salle salle;

    // ✅ AJOUT : Relation directe vers l'Espace
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "espace_id", nullable = true)
    private Espace espace;

    @OneToMany(mappedBy = "reservation", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    @JsonIgnore
    private List<Paiement> paiements = new ArrayList<>();

    @OneToOne(mappedBy = "reservation", cascade = CascadeType.ALL, orphanRemoval = false, fetch = FetchType.LAZY)
    private Contrat contrat;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "reservation_services",
            joinColumns = @JoinColumn(name = "reservation_id"),
            inverseJoinColumns = @JoinColumn(name = "service_id")
    )
    @Builder.Default
    private List<Service> services = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cree_par", nullable = false)
    private User creePar;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "reservation_tarifications",
            joinColumns = @JoinColumn(name = "reservation_id"),
            inverseJoinColumns = @JoinColumn(name = "tarification_id")
    )
    @Builder.Default
    private List<TarificationSalle> tarificationsAppliquees = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "modifie_par")
    private User modifiePar;

    @Column(nullable = false, updatable = false)
    private LocalDateTime dateCreation;

    private LocalDateTime dateModification;
}